examples/generative/molecule_generation.py

```diff
@@ -0,0 +1,493 @@
+
+        self.train_total_loss_tracker.update_state(total_loss)
+        self.train_reconstruction_loss_tracker.update_state(reconstruction_loss)
+        self.train_kl_loss_tracker.update_state(kl_loss)
+        self.train_pp_loss_tracker.update_state(pp_loss)
+        return {
+            "loss": self.train_total_loss_tracker.result(),
+            "reconstruction_loss": self.train_reconstruction_loss_tracker.result(),
+            "kl_loss": self.train_kl_loss_tracker.result(),
+            "property_prediction_loss": self.train_pp_loss_tracker.result(),
+        }
+
+    def test_step(self, data):
+        smile, sas = data
+
+        z_mean, z_log_var, z, reconstruction, property_prediction = self(
+            smile, training=False
+        )
+
+        reconstruction_loss = tf.reduce_mean(
+            tf.reduce_sum(
+                tf.keras.losses.categorical_crossentropy(smile, reconstruction),
+                axis=(1),
+            )
+        )
+
+        kl_loss = -0.5 * tf.reduce_sum(
+            1 + z_log_var - tf.square(z_mean) - tf.exp(z_log_var), 1
+        )
+        kl_loss = tf.reduce_mean(kl_loss)
+
+        pp_loss = tf.keras.losses.mean_squared_error(sas, property_prediction)
+        pp_loss = tf.reduce_sum(
+            tf.keras.losses.mean_squared_error(sas, property_prediction)
+        )
+
+        total_loss = kl_loss + pp_loss + reconstruction_loss
+
+        self.val_total_loss_tracker.update_state(total_loss)
+        self.val_reconstruction_loss_tracker.update_state(reconstruction_loss)
+        self.val_kl_loss_tracker.update_state(kl_loss)
+        self.val_pp_loss_tracker.update_state(pp_loss)
+
+        return {
+            "loss": self.val_total_loss_tracker.result(),
+            "reconstruction_loss": self.val_reconstruction_loss_tracker.result(),
+            "kl_loss": self.val_kl_loss_tracker.result(),
+            "property_prediction_loss": self.val_pp_loss_tracker.result(),
+        }
+
+    def call(self, x):
+        z_mean, log_var = self.encoder(x)
+        z = Sampling()([z_mean, log_var])
+
+        reconstruction = self.decoder(z)
+
+        property_prediction = self.property_prediction_layer(z_mean)
+
+        return z_mean, log_var, z, reconstruction, property_prediction
+
+
+"""
+## Model Training
+"""
+
+train_loader = DataGenerator(
+    data=train_df[:5000], mapping=SMILE_to_index, max_len=MAX_LEN, batch_size=BATCH_SIZE
+)
+
+validation_loader = DataGenerator(
+    data=test_df[:5000], mapping=SMILE_to_index, max_len=MAX_LEN, batch_size=BATCH_SIZE
+)
+
+vae_optimizer = tf.keras.optimizers.Adam(learning_rate=VAE_LR)
+
+encoder = Encoder(latent_dim)
+decoder = Decoder(latent_dim, max_length=MAX_LEN, charset_length=len(SMILE_CHARSET))
+
+model = MoleculeGenerator(encoder, decoder, MAX_MOLSIZE)
+
+model.compile(vae_optimizer)
+history = model.fit(train_loader, epochs=EPOCHS, validation_data=validation_loader)
```

Comment:
Does using `num_workers` improve performance when running on a multicore machine?

Reply:
I removed the data loader segment because it consumed too many lines of code. But I am not totally sure whether num_workers would improve model performance.